ReportUpdater.php

```diff
@@ -22,97 +67,137 @@ public function __construct() {
-				$this->api->setText( $info['Report'], $output, 1 );
-			} else {
-				// Update complete page
-				$this->api->setText( $info['Report'], $output );
-			}
-			logToFile( 'Finished processing: ' . $info['Name'] );
+
+			// Generate and save the report.
+			$this->processProject( $project, $projectConfig );
+
+			wfLogToFile( 'Finished processing: ' . $projectConfig['Name'] );
+		}
+
+		die();
+
+		// Update index page.
+		$this->updateIndex();
+	}
+
+	/**
+	 * Process an individual WikiProject and update its popular pages report.
+	 *
+	 * @param  string $project
+	 * @param  array $config As specified in the on-wiki JSON config.
+	 * @return string Date when report completed.
+	 */
+	private function processProject( $project, $config ) {
+		// Returns { 'title' => ['class'=>'', 'importance'=>''],...}
+		$pages = $this->api->getProjectPages( $config['Name'] );
+
+		if ( empty( $pages ) ) {
+			return;
+		}
+
+		if ( count( $pages ) > 1000000 ) { // See T164178
+			wfLogToFile( 'Error: ' . $project . ' is too large. Skipping.' );
+			return;
+		}
+
+		$pageviews = $this->api->getMonthlyPageviews(
+			array_keys( $pages ),
+			date( 'Ymd00', $this->start ),
+			date( 'Ymd00', $this->end )
+		);
+
+		// Compute total views for the month
+		$totalViews = array_sum( array_values( $pageviews ) );
+
+		// Truncate to configured limit.
+		$pageviews = array_slice( $pageviews, 0, $config['Limit'], true );
+
+		// Populate new reported pages.
+		foreach ( $pageviews as $title => $views ) {
+			$pageviews[$title] = array_merge( $pages[$title], [
+				'views' => $views,
+				'avgViews' => floor(
+					$views / ( floor( ( $this->end - $this->start ) / ( 60 * 60 * 24 ) ) )
+				),
+			] );
+
+			// Attempt to reduce memory usage, probably unneeded.
+			unset( $pages[$title] );
+		}
+
+		// Another attempt to reduce memory usage, again probably unneeded.
+		$pages = null;
+
+		$hasListSection = $this->api->hasLeadSection( $config['Report'] );
+
+		// Generate and return wikitext.
+		$output = $this->twig->render( 'report.wikitext.twig', [
+			'hasListSection' => $hasListSection,
+			'wiki' => $this->wiki,
+			'start' => $this->start,
+			'end' => $this->end,
+			'project' => $project,
+			'pages' => $pageviews,
+			'totalViews' => $totalViews,
+			'config' => $this->api->getAssessmentConfig(),
+		] );
+
+		$this->api->setText( $config['Report'], $output, (int)$hasListSection );
+	}
+
+	/**
+	 * Update index page for the bot, listing each WikiProject, its report
+	 * and when it was last updated.
+	 */
+	private function updateIndex() {
+		$projectsConfig = $this->api->getJSONConfig();
+
+		foreach ( $projectsConfig as $project => &$data ) {
+			$data['Updated'] = $this->api->getBotLastEditDate( $project );
+		}
+
+		// Generate and return wikitext.
+		$output = $this->twig->render( 'index.wikitext.twig', [
+			'projects' => $projectsConfig,
+		] );
+
+		$this->api->setText( 'User:Community Tech bot/Popular pages', $output );
```

Comment:
This would need to change to point to the WP: namespace page on the local wiki. We should have this as a message string now, I think.